cpp/src/arrow/array/builder_adaptive.h

```diff
@@ -25,9 +25,24 @@ namespace arrow {
 
 namespace internal {
 
+template <uint8_t size>
+constexpr uint64_t ExpandIntSizeMask() {
+  return ~static_cast<uint64_t>(0) << size * CHAR_BIT;
+}
+
+template <>
+constexpr uint64_t ExpandIntSizeMask<sizeof(int64_t)>() {
+  return 0;
+}
+
+template <uint8_t size>
+constexpr uint64_t ExpandIntSizeAddend() {
+  return static_cast<uint64_t>(1) << (size * CHAR_BIT - 1);
+}
+
 class ARROW_EXPORT AdaptiveIntBuilderBase : public ArrayBuilder {
  public:
-  explicit AdaptiveIntBuilderBase(MemoryPool* pool);
+  AdaptiveIntBuilderBase(const std::shared_ptr<DataType>& type, MemoryPool* pool);
```

Comment:
Can you add a docstring? It's not immediately obvious what the datatype should be.